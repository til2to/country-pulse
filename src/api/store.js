import { configureStore } from '@reduxjs/toolkit';
import continentSlice from './continentSlice';
import countrySlice from './countrySlice';

const store = configureStore({
  reducer: {
    continent: continentSlice,
    country: countrySlice,
  },
});

export default store;
