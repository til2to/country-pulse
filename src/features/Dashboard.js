import React, { useEffect } from 'react';
import africa from '../assets/africa.png';
import asia from '../assets/asia.png';
import europe from '../assets/europe.png';
import northAmerica from '../assets/north-america.png';
import oceania from '../assets/oceania.png';
import southAmerica from '../assets/south-america.png';
import ContinentCard from '../components/ContinentCard';

const Dashboard = () => {
  const links = [
    {
      name: 'Africa',
      path: '/continent/africa',
      image: africa,
    },
    {
      name: 'Asia',
      path: '/continent/asia',
      image: asia,
    },
    {
      name: 'Europe',
      path: '/continent/europe',
      image: europe,
    },
    {
      name: 'North America',
      path: '/continent/north america',
      image: northAmerica,
    },
    {
      name: 'South America',
      path: '/continent/south america',
      image: southAmerica,
    },
    {
      name: 'Oceania',
      path: '/continent/oceania',
      image: oceania,
    },
  ];

  useEffect(() => {
    window.scrollTo(0, 0);
  }, []);

  return (
    <section className="p-6 grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-6">
      {links.map((link) => (
        <ContinentCard key={link.name} link={link} />
      ))}
    </section>
  );
};
export default Dashboard;
