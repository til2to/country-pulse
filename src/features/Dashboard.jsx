import React from 'react';
import SearchFilter from '../components/SearchFilter';

const Dashboard = () => (
  <>
    <SearchFilter />
  </>
);
export default Dashboard;
