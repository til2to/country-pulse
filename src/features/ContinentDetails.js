import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { fetchCountries } from '../api/continentSlice';
import CountryCard from '../components/CountryCard';
import SearchFilter from '../components/SearchFilter';

const ContinentDetails = () => {
  const { countries, loading, error } = useSelector((state) => state.continent);
  const { continent } = useParams();
  const [query, setQuery] = useState('');
  const dispatch = useDispatch();

  const getContent = () => {
    let content = null;

    if (loading) {
      content = (
        <div className="h-[18rem] flex justify-center items-center">
          <p className="text-center text-2xl">Getting data...</p>
        </div>
      );
    } else if (!loading) {
      const filteredCountries = countries?.filter((country) => {
        const nameMatchesQuery = country.name.common
          .toLowerCase()
          .includes(query.toLowerCase());

        const queryIsEmpty = query === '';
        return queryIsEmpty || nameMatchesQuery;
      });

      content = (
        <>
          {filteredCountries.length === 0 ? (
            <p className="text-center text-2xl">No countries found</p>
          ) : (
            filteredCountries.map((country) => (
              <CountryCard key={country.name.common} country={country} />
            ))
          )}
        </>
      );
    } else if (error) {
      content = <p>{error}</p>;
    }

    return content;
  };

  useEffect(() => {
    dispatch(fetchCountries(continent));
  }, [dispatch, continent]);

  return (
    <>
      <div className="bg-slate-700 text-slate-100 w-full h-[14rem] shadow pl-3 py-5 flex items-center justify-center gap-x-4">
        <h3 className="text-2xl font-bold">
          {countries.length}
          {' '}
          Countries
        </h3>
      </div>

      <SearchFilter query={query} setQuery={setQuery} />

      <section className="px-8 py-5 lg:px-10 grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-10 lg:gap-16">
        {getContent()}
      </section>
    </>
  );
};
export default ContinentDetails;
