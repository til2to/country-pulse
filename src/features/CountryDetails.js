import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { fetchCountry } from '../api/countrySlice';

const CountryDetails = () => {
  const {
    country: data,
    loading,
    error,
  } = useSelector((state) => state.country);
  const { name } = useParams();
  const dispatch = useDispatch();
  const [country, setCountry] = useState(null);

  const getContent = () => {
    let content = null;

    if (loading) {
      content = (
        <div className="h-screen flex justify-center items-center">
          <p className="text-center text-2xl">Getting data...</p>
        </div>
      );
    } else if (!loading && country) {
      content = (
        <section className="h-[35rem] grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 mt-4 sm:mt-0">
          <div className="grid place-content-center px-4 mb-4 sm:mb-0">
            <img
              src={country.flags.png}
              alt={country.name.common}
              className="shadow-md w-full lg:h-[20rem] object-cover"
            />
          </div>
          <div className="md:grid place-content-center pl-16 sm:pl-0">
            <p className="text-xl mb-3">
              <span className="font-bold">Native Name: </span>
              {
                country.name.nativeName[Object.keys(country.name.nativeName)[0]]
                  .official
              }
            </p>
            <p className="text-xl mb-3">
              <span className="font-bold">Population: </span>
              {country.population.toLocaleString()}
            </p>
            <p className="text-xl mb-3">
              <span className="font-bold">Area: </span>
              {country.area.toLocaleString()}
              {' '}
              <span>
                km
                <sup>2</sup>
              </span>
            </p>
            <p className="text-xl mb-3">
              <span className="font-bold">Geolocation: </span>
              Latitude
              {' '}
              {country.latlng[0]}
              , Longitude
              {' '}
              {country.latlng[1]}
            </p>
            <p className="text-xl mb-3">
              <span className="font-bold">Sub Region: </span>
              {country.subregion}
            </p>
            <p className="text-xl mb-3">
              <span className="font-bold">Region: </span>
              {country.region}
            </p>
            <p className="text-xl mb-3">
              <span className="font-bold">Capital: </span>
              {country.capital[0]}
            </p>
            <p className="text-xl mb-3">
              <span className="font-bold">Currency: </span>
              {country.currencies[Object.keys(country.currencies)[0]].name}
              {' - '}

              <span className="font-bold">
                {country.currencies[Object.keys(country.currencies)[0]].symbol}
              </span>
            </p>
          </div>
        </section>
      );
    } else if (error) {
      content = <p className="text-center text-2xl">{error}</p>;
    }

    return content;
  };

  useEffect(() => {
    dispatch(fetchCountry(name));
  }, [dispatch, name]);

  useEffect(() => {
    setCountry(data[0]);
  }, [data]);

  return <>{getContent()}</>;
};
export default CountryDetails;
