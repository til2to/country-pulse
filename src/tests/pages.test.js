import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../api/store';
import ContinentDetails from '../features/ContinentDetails';
import CountryDetails from '../features/CountryDetails';
import Dashboard from '../features/Dashboard';

describe('UI Tests', () => {
  it('should render Dashboard', () => {
    const tree = render(
      <BrowserRouter>
        <Dashboard />
      </BrowserRouter>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should render Continent Details', () => {
    const tree = render(
      <Provider store={store}>
        <BrowserRouter>
          <ContinentDetails />
        </BrowserRouter>
      </Provider>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should render Country Details', () => {
    const tree = render(
      <Provider store={store}>
        <BrowserRouter>
          <CountryDetails />
        </BrowserRouter>
      </Provider>,
    );
    expect(tree).toMatchSnapshot();
  });
});
