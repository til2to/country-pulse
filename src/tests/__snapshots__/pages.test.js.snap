// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI Tests should render Continent Details 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bg-slate-700 text-slate-100 w-full h-[14rem] shadow pl-3 py-5 flex items-center justify-center gap-x-4"
      >
        <h3
          class="text-2xl font-bold"
        >
          0
           
          Countries
        </h3>
      </div>
      <div
        class="p-2 my-2 lg:w-1/2 mx-auto"
      >
        <input
          class="w-full p-3 rounded-md border shadow-md shadow-slate-300 text-lg focus:outline-none"
          placeholder="Search for a country..."
          type="search"
          value=""
        />
      </div>
      <section
        class="px-8 py-5 lg:px-10 grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-10 lg:gap-16"
      >
        <div
          class="h-[18rem] flex justify-center items-center"
        >
          <p
            class="text-center text-2xl"
          >
            Getting data...
          </p>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-slate-700 text-slate-100 w-full h-[14rem] shadow pl-3 py-5 flex items-center justify-center gap-x-4"
    >
      <h3
        class="text-2xl font-bold"
      >
        0
         
        Countries
      </h3>
    </div>
    <div
      class="p-2 my-2 lg:w-1/2 mx-auto"
    >
      <input
        class="w-full p-3 rounded-md border shadow-md shadow-slate-300 text-lg focus:outline-none"
        placeholder="Search for a country..."
        type="search"
        value=""
      />
    </div>
    <section
      class="px-8 py-5 lg:px-10 grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-10 lg:gap-16"
    >
      <div
        class="h-[18rem] flex justify-center items-center"
      >
        <p
          class="text-center text-2xl"
        >
          Getting data...
        </p>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`UI Tests should render Country Details 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="h-screen flex justify-center items-center"
      >
        <p
          class="text-center text-2xl"
        >
          Getting data...
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="h-screen flex justify-center items-center"
    >
      <p
        class="text-center text-2xl"
      >
        Getting data...
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`UI Tests should render Dashboard 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="p-6 grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-6"
      >
        <div>
          <a
            class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
            href="/continent/africa"
          >
            Africa
            <div
              class="w-[10rem]"
            >
              <img
                alt="Africa"
                class="object-contain"
                src="africa.png"
              />
            </div>
          </a>
        </div>
        <div>
          <a
            class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
            href="/continent/asia"
          >
            Asia
            <div
              class="w-[10rem]"
            >
              <img
                alt="Asia"
                class="object-contain"
                src="asia.png"
              />
            </div>
          </a>
        </div>
        <div>
          <a
            class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
            href="/continent/europe"
          >
            Europe
            <div
              class="w-[10rem]"
            >
              <img
                alt="Europe"
                class="object-contain"
                src="europe.png"
              />
            </div>
          </a>
        </div>
        <div>
          <a
            class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
            href="/continent/north america"
          >
            North America
            <div
              class="w-[10rem]"
            >
              <img
                alt="North America"
                class="object-contain"
                src="north-america.png"
              />
            </div>
          </a>
        </div>
        <div>
          <a
            class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
            href="/continent/south america"
          >
            South America
            <div
              class="w-[10rem]"
            >
              <img
                alt="South America"
                class="object-contain"
                src="south-america.png"
              />
            </div>
          </a>
        </div>
        <div>
          <a
            class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
            href="/continent/oceania"
          >
            Oceania
            <div
              class="w-[10rem]"
            >
              <img
                alt="Oceania"
                class="object-contain"
                src="oceania.png"
              />
            </div>
          </a>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="p-6 grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-6"
    >
      <div>
        <a
          class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
          href="/continent/africa"
        >
          Africa
          <div
            class="w-[10rem]"
          >
            <img
              alt="Africa"
              class="object-contain"
              src="africa.png"
            />
          </div>
        </a>
      </div>
      <div>
        <a
          class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
          href="/continent/asia"
        >
          Asia
          <div
            class="w-[10rem]"
          >
            <img
              alt="Asia"
              class="object-contain"
              src="asia.png"
            />
          </div>
        </a>
      </div>
      <div>
        <a
          class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
          href="/continent/europe"
        >
          Europe
          <div
            class="w-[10rem]"
          >
            <img
              alt="Europe"
              class="object-contain"
              src="europe.png"
            />
          </div>
        </a>
      </div>
      <div>
        <a
          class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
          href="/continent/north america"
        >
          North America
          <div
            class="w-[10rem]"
          >
            <img
              alt="North America"
              class="object-contain"
              src="north-america.png"
            />
          </div>
        </a>
      </div>
      <div>
        <a
          class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
          href="/continent/south america"
        >
          South America
          <div
            class="w-[10rem]"
          >
            <img
              alt="South America"
              class="object-contain"
              src="south-america.png"
            />
          </div>
        </a>
      </div>
      <div>
        <a
          class="w-full h-[14rem] continent-card text-2xl font-semibold pl-3 py-5 rounded-md flex items-center justify-center gap-x-4"
          href="/continent/oceania"
        >
          Oceania
          <div
            class="w-[10rem]"
          >
            <img
              alt="Oceania"
              class="object-contain"
              src="oceania.png"
            />
          </div>
        </a>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
